### PR TITLE
Refactored LLM class.

### PR DESCRIPTION
**Pull request by SmartManoj**
#### Summary for a single change.
```
- model if model is not None else llm_config.model
+  model or llm_config.model
```
This PR refactors the code for setting the `model` attribute to use a more concise and Pythonic approach. The new implementation leverages the `or` operator to simplify default value handling, ensuring clarity and readability.

#### Changes Made
- Updated the assignment of `model` from:
  ```python
  model if model is not None else llm_config.model
  ```
  to:
  ```python
  model or llm_config.model
  ```

#### Reason for Change
- The previous implementation explicitly checks if `model` is not `None`, which, while clear, is more verbose than necessary.
- The new implementation using `or` is more concise and achieves the same result, given that falsey values (e.g., `None`, `0`, `False`, `''`) are not passed in this context.

#### Benefits
- **Simplicity**: The new code is shorter and easier to read.
- **Pythonic**: Using the `or` operator is a common Python idiom for handling default values.

#### Testing
- Verified that `model` is correctly assigned to `llm_config.model` when `model` is not provided.
- Ensured existing tests pass and ~~added new tests to cover edge cases.~~  (LLM Generated summary 😬 as it is hard to convince this)

#### Example
Before:
```python
self.model = model if model is not None else llm_config.model
```
After:
```python
self.model = model or llm_config.model
```

With this change, the codebase becomes cleaner and maintains the same functionality.

### CODE REVIEW

File: opendevin/llm/llm.py [hunk 1]
```diff
@@ -86,49 +86,28 @@ def __init__(
...
-        max_output_tokens = (
-            max_output_tokens
-            if max_output_tokens is not None
-            else llm_config.max_output_tokens
-        )
-        metrics = metrics if metrics is not None else Metrics()
+        # not used in the current version
+        self.custom_llm_provider = custom_llm_provider or llm_config.custom_llm_provider
+        self.max_input_tokens = max_input_tokens or llm_config.max_input_tokens
+        self.max_output_tokens = max_output_tokens or llm_config.max_output_tokens
```
Comment (enyst) [inline opendevin/llm/llm.py:106]: Ditto, I don't know if 0 has meaning here, or will it have, but they're numeric params. So I'd assume if the user set 0, they reasonably expect us to send 0.

**Review by enyst — CHANGES_REQUESTED** [review]
> and added new tests to cover edge cases

I don't see any new tests 😅 

I think these changes can simplify some cases, like `model`, which I don't know, I don't expect to be 'falsey'. But others seem to have valid use cases today that will break, because the 'falsey' values are meaningful.

File: opendevin/llm/llm.py [hunk 2]
```diff
@@ -86,49 +86,28 @@ def __init__(
...
-            else llm_config.max_input_tokens
-        )
-        max_output_tokens = (
-            max_output_tokens
-            if max_output_tokens is not None
-            else llm_config.max_output_tokens
-        )
-        metrics = metrics if metrics is not None else Metrics()
+        # not used in the current version
+        self.custom_llm_provider = custom_llm_provider or llm_config.custom_llm_provider
```
Comment (enyst) [inline opendevin/llm/llm.py:104]:
```suggestion
        self.custom_llm_provider = (
            custom_llm_provider
            if custom_llm_provider is not None
            else llm_config.custom_llm_provider
        )
```
This is probably similar with `base_url`? I think '' might be a valid value that works for openai.

**Comment by tobitege** [comment]
I'm sorry to say, but I think most of these changes shouldn't be merged (strings and numerics are falsey values to the "or" parameter), unless all changes are covered by unit tests to see the differences in outcomes.
@enyst already summed it up. And if something is still going as None to litellm, let it deal with it.
E.g. temperature of 0 is definitely a valid and often wanted value.

File: opendevin/llm/llm.py [hunk 3]
```diff
@@ -86,49 +86,28 @@ def __init__(
...
         )
         retry_max_wait = (
             retry_max_wait if retry_max_wait is not None else llm_config.retry_max_wait
         )
-        llm_timeout = llm_timeout if llm_timeout is not None else llm_config.timeout
-        llm_temperature = (
-            llm_temperature if llm_temperature is not None else llm_config.temperature
-        )
+        self.llm_timeout = llm_timeout or llm_config.timeout
+        llm_temperature = llm_temperature or llm_config.temperature
```
[the remark this answers is not in the document]
Comment (SmartManoj) [reply opendevin/llm/llm.py:101]: default value for temperature is 0.

File: opendevin/llm/llm.py [hunk 4]
```diff
@@ -86,49 +86,28 @@ def __init__(
         """
         if llm_config is None:
             llm_config = config.llm
-        model = model if model is not None else llm_config.model
-        api_key = api_key if api_key is not None else llm_config.api_key
-        base_url = base_url if base_url is not None else llm_config.base_url
-        api_version = api_version if api_version is not None else llm_config.api_version
+        self.model_name = model or llm_config.model
+        self.api_key = api_key or llm_config.api_key
+        self.base_url = base_url or llm_config.base_url
```
[the remark this answers is not in the document]
Comment (SmartManoj) [reply opendevin/llm/llm.py:91]: It also works with default value (None). Right?

File: opendevin/llm/llm.py (hunk 1 above)
Comment (SmartManoj) [reply opendevin/llm/llm.py:106]: If they pass `None`?
https://github.com/OpenDevin/OpenDevin/blob/6c741372ffff71a98f2f03bcc0dc4841cd444871/opendevin/llm/llm.py#L131-L137

File: opendevin/llm/llm.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tobitege) [reply opendevin/llm/llm.py:101]: Here the issue is not the default value itself, but that the "or" would result in potentially a different value if "llm_temperature" were 0, i.e. llm_config.temperature.
The "is not None" prevents this sideeffect.

File: opendevin/llm/llm.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (tobitege) [reply opendevin/llm/llm.py:91]: Here as wel: if base_url was intentionally left empty, the "or" would switch to potentially a different value.

[the remark this answers is not in the document]
Comment (SmartManoj) [reply opendevin/llm/llm.py:91]: So, the side affect?

File: opendevin/llm/llm.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (SmartManoj) [reply opendevin/llm/llm.py:101]: 0 or 0 => 0

File: opendevin/llm/llm.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (tobitege) [reply opendevin/llm/llm.py:91]: If LLM class is initialized with the parameter being None (the default for most __init__ params), changing the `is not None else ` to simply `or` for any numeric parameter being `0` or string parameter being empty would result in effectively using the corresponding value from "llm_config". 
This would potentially mix values even though empty/0 were intentional values.

File: opendevin/llm/llm.py (hunk 1 above)
Comment (tobitege) [reply opendevin/llm/llm.py:106]: > If they pass `None`?
> 
> https://github.com/OpenDevin/OpenDevin/blob/6c741372ffff71a98f2f03bcc0dc4841cd444871/opendevin/llm/llm.py#L131-L137

`max_input_tokens` and `max_output_tokens` are already dealt with at that point:
https://github.com/OpenDevin/OpenDevin/blob/90ec0095dfe72e30114e093f2dc27c6bc75e8140/opendevin/llm/llm.py#L110-L119

Comment (SmartManoj) [reply opendevin/llm/llm.py:106]: If the user set `None`, they also reasonably expect us to send `None`.

Comment (tobitege) [reply opendevin/llm/llm.py:106]: > If the user set `None`, they also reasonably expect us to send `None`.

I'm not clear what you propose to change here?

_If_ the model returns actual values for min/max tokens, those values are hard facts and imho have precedence over any user-made config/param, even if None (talking about these "if self.min/max_output_tokens is None:" blocks).

Comment (tobitege) [reply opendevin/llm/llm.py:106]: In a cursory search in the codebase I couldn't find that e.g. `max_tokens` or the min/max_output_tokens are actually used yet for other purposes than in the config.
Maybe these code locations should be revisited once that is the case?

Comment (SmartManoj) [reply opendevin/llm/llm.py:106]: Currently, if the user gives None, the value is changed. 
0 is not a valid value.

File: opendevin/llm/llm.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (SmartManoj) [reply opendevin/llm/llm.py:91]: Here, it will change to `None`. And no, side effect.

File: opendevin/llm/llm.py (hunk 1 above)
Comment (SmartManoj) [reply opendevin/llm/llm.py:106]: It is used here.
https://github.com/OpenDevin/OpenDevin/blob/6c741372ffff71a98f2f03bcc0dc4841cd444871/opendevin/llm/llm.py#L146

Comment (SmartManoj) [reply opendevin/llm/llm.py:106]: Why should one set it to zero?

Comment (tobitege) [reply opendevin/llm/llm.py:106]: Just to reset the context of what we are discussing here in class `LLM`:
the variables that are checked here are values coming from the codebase, e.g. if an agent would set a specific value - for whatever reason. Thus if they are not None, they are used and not the value coming from the `llm_config`. The code would rely the LLM class to obey this to make the deterministic and not infuse an unknown value from the _user_.
When saying _user_ it is what is set in the `config.toml`'s `llm` section, in `llm_config`.
The code here is mostly to determine **precedence** of values, runtime parameter vs. config.
And the `or` operator introduces behavioral change with regard to `falsey` values like 0 or empty string.
We are _not_ checking each and every value for validity coming from the `llm` section.
The _user_ of OpenDevin is responsible to configure sensible values.

Comment (tobitege) [reply opendevin/llm/llm.py:106]: That latest commit is certainly an improvement, yes, but doesn't change my post above.

File: opendevin/llm/llm.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tobitege) [reply opendevin/llm/llm.py:101]: `No. 0 or 0 => 0`
Huh?
If **parameter** `llm_temperature` was 0 (via code!), the `or` would assign `llm_config.temperature` (user value), which could be e.g. 0.2 and thus not the expected value as requested by code.

**Comment by tobitege** [comment]
Just as an additional explanation for my view points in above comments (worded by gpt4o):

_When explaining why changing the code from using `if` to using `or` might introduce issues, it's important to highlight the differences in how these constructs handle falsy values. Here’s a detailed explanation:_

### Original Code:
```python
llm_temperature = (
    llm_temperature if llm_temperature is not None else llm_config.temperature
)
```

### Modified Code:
```python
llm_temperature = llm_temperature or llm_config.temperature
```

### Key Points to Explain:

1. **Handling of Falsy Values**:
   - **Original Code**: The expression `llm_temperature if llm_temperature is not None else llm_config.temperature` specifically checks if `llm_temperature` is `None`. If it's `None`, it uses `llm_config.temperature`; otherwise, it uses the value of `llm_temperature`. This means it only substitutes the default value if `llm_temperature` is explicitly `None`.
   - **Modified Code**: The expression `llm_temperature or llm_config.temperature` checks if `llm_temperature` is falsy (i.e., evaluates to `False` in a boolean context). In Python, values like `0`, `False`, `''` (empty string), and `[]` (empty list) are all considered falsy. Therefore, if `llm_temperature` is any of these values, `llm_config.temperature` will be used instead.

2. **Practical Implications**:
   - If `llm_temperature` is intended to accept values that could be falsy but still valid (e.g., `0` or `0.0` as a valid temperature), using `or` would incorrectly substitute these values with `llm_config.temperature`. This could lead to unexpected behavior and bugs.
   - The original code ensures that only a `None` value triggers the fallback to the configuration value, preserving the intended meaning of valid falsy values.

### Example Scenario:
Suppose `llm_temperature` can be `0.0`, which is a valid and meaningful value, but also falsy in a boolean context.

- **Original Code**: 
  ```python
  llm_temperature = 0.0  # Would remain 0.0
  ```
- **Modified Code**: 
  ```python
  llm_temperature = 0.0 or llm_config.temperature  # Would become llm_config.temperature, which is incorrect
  ```

### Conclusion:
By changing the code to use the `or` operator, you risk incorrectly handling valid falsy values. The original code's explicit check for `None` ensures that only an unset (`None`) value for `llm_temperature` triggers the fallback, preserving all other valid values, including falsy ones. This distinction is crucial for maintaining the correct and expected behavior of your function.

**Comment by SmartManoj** [comment]
image
FYI, only these values are directly passed to the config now.

**Comment by tobitege** [comment]
> image FYI, only these values are directly passed to the LLM now.

That is besides the point. The current code tries to cover most runtime parameters, regardless of currently being used or not.

**Comment by SmartManoj** [comment]
> only these values are directly passed to the LLM now.

Are the remaining params necessary?

**Comment by enyst** [comment]
Just a few notes on the discussion, and I'll follow-up on the question.

Re: precedence.
> The code here is mostly to determine **precedence** of values, runtime parameter vs. config.

I agree, this is a matter of precedence: explicit params > env vars > user config file > app default config. Similar kind of argument like [here](https://stackoverflow.com/questions/11077223/what-order-of-reading-configuration-values#:~:text=This%20can%20be%20changed%20with%20a%20local%20config,command%20line%20params%20will%20take%20precedence%20over%20it.).

Re: None.
> If the user set `None`, they also reasonably expect us to send `None`.

It's the default value, so it's equivalent with no value set. I'd say it's a reasonable expectation in python that the function they called behaves as if they set no value in their code. Special cases are possible I guess, but why complicate things? Other languages do different stuff on optional arguments like this or don't support them at all.

**Comment by enyst** [comment]
> @enyst Are the remaining params necessary?

Some are used in evals. It's true it's not the same, at least since we introduced llm_config customization: they're now used by defining [multiple custom llm_configs](https://github.com/OpenDevin/OpenDevin/blob/main/evaluation/humanevalfix/README.md). Before that, one had to find other ways to override.

More to the point here, though: this PR is not proposing to remove them, but to change them when they have meaningful user-set values. And that seems that it would break not only common practice or our old use cases, but also OpenAI compatibility.

These parameters are the parameters defined by OpenAI API for the completion call. Think of all the ways that matters today. Open LLMs have a compatible API at least in part; applications like LMStudio offer a compatible API afaik; same for other proxies. LiteLLM has an explicit documented goal to keep OpenAI compatibility for other LLMs as much as possible. Research papers test on similar API, etc. Thousands (or more?) of lines of code out there use it.

Currently, someone who wants to adapt their agent to work with opendevin doesn't need to change their LLM code at all or almost: set it up to use our class and it just works. We haven't documented much, if at all, this drop-in compatibility feature, but it's real. [1] 

More importantly, they shouldn't need to debug forever why `temperature=0` that they send to our LLM doesn't behave the same way it did when they sent it to OpenAI API. Or directly to our own completion method, even.

(that said, I still think "model" is harmless 🤷. Happy to be proven otherwise.)

Side note:
As an example of how it matters, this CodeAct line: https://github.com/OpenDevin/OpenDevin/blob/90ec0095dfe72e30114e093f2dc27c6bc75e8140/agenthub/codeact_agent/codeact_agent.py#L226
We don't have a `stop` variable. I didn't even have a clue there was such thing. But OpenAI does have a [`stop` parameter](https://platform.openai.com/docs/api-reference/chat/create#chat-create-stop), and CodeAct completion() call just works.
Please note that it also sets temperature=0.0.


[1] I think we should keep the signature as clearly as possible as "drop-in replacement" and that's why I'm nitpicking even the use of a different name ( "do_completion") for the completion function, though it's not a big deal, but it's [fixed here](https://github.com/OpenDevin/OpenDevin/commit/c234d82387349dbb1352e4ae22cc2cda0f211479).

**Comment by SmartManoj** [comment]
> More importantly, they shouldn't need to debug forever why `temperature=0` that they send to our LLM doesn't behave the same way it did when they sent it to OpenAI API. Or directly to our own completion method, even.

https://github.com/OpenDevin/OpenDevin/pull/2397#discussion_r1635742056

**Comment by enyst** [comment]
@SmartManoj the comment you linked says:
> default value for temperature is 0. So, no effect unless we change it. Will we?

I doubt we will change it. On the contrary, it seems a well known value and appropriate for the precision we need as a software tool. That doesn't mean I can predict the future. 😅 It means I don't see it.

That misses the point though. It can still happen, but also the user can set their own in config.toml. So the user can have 0.5 because they liked how Claude worked with it for a novel, let's say. So this PR overrides the parameter with 0.5.

Not only that, correct me if wrong: 
1) it sends `0.0` to the API when the agent calls `llm.completion(temperature=0.0)`, 
2) but it sends `0.5` when the agent instantiates the LLM class with `temperature=0.0` then calls `llm.completion()`.

From the perspective of an agent developer, this is completely unpredictable - and wrong - behavior.

I suggest to focus this PR on those params where the syntax change doesn't change behavior. Refactoring means no behavior change, after all.

**Comment by SmartManoj** [comment]
> 2. but it sends `0.5` when the agent instantiates the LLM class with `temperature=0.0` then calls `llm.completion()`.

Why should the agent not respect the user's temperature config?

**Review by neubig — CHANGES_REQUESTED** [review]
I think we should probably not do this change. `x or y` is not always equivalent to `x if x is not None else y`.

Also, if we are talking about Pythonic code, the second rule in [The Zen of Python](https://peps.python.org/pep-0020/), is "Explicit is better than Implicit". I would prefer an explicit check for `None` rather than an implicit one, especially when the implicit behavior is less predictable.

**Comment by neubig** [comment]
I am going to close this, since it seems that most people would prefer to not introduce the change, but we can re-open if there are contrary opinions!

**Comment by SmartManoj** [comment]
While I understand your concerns, I'd like to point out that simplicity is key in Pythonic code. According to the Zen of Python, "Simple is better than complex." The x or y approach is simpler and more readable, which aligns with this principle. It reduces unnecessary verbosity and keeps the code clean.

Furthermore, predictability can be maintained through consistent use and proper documentation. If there is still a disagreement, we can certainly proceed with a vote and decide by majority.

**Comment by tobitege** [comment]
It is not about simplicity or better readability here.
The use of the "or" operator in this PR changes prioritization of values, if e.g. a variable is None, changing the logic and outcome of the value assignments.

**Comment by SmartManoj** [comment]
> changing the logic and outcome of the value assignments.

 changing the logic but same outcome

**Comment by tobitege** [comment]
> > changing the logic and outcome of the value assignments.
> 
> changing the logic but same outcome

Wrong.

**Comment by SmartManoj** [comment]
> Wrong.

@tobitege Could you please explain?

**Comment by tobitege** [comment]
> > Wrong.
> 
> @tobitege Could you please explain?

I've given all explanations and arguments above, but you seem to refuse to take them into account.

**Comment by SmartManoj** [comment]
> if e.g. a variable is None, changing the logic and outcome of the value assignments.

@tobitege could you point out a variable because so many variables are involved here?
